knimin/lib/data_access.py

```diff
@@ -3894,6 +3894,103 @@ def delete_shotgun_index_aliquot(self, shotgun_index_aliquot_id):
             TRN.add(sql, [shotgun_index_aliquot_id])
             TRN.execute()
 
+    def generate_i5_i7_indexes(self, idx_tech, num_samples):
+        """Generates pairs of (i5_id, i7_id)
+
+        Parameters
+        ----------
+        idx_tech : str
+            The index technology we want to use
+        num_samples : int
+            The number of pairs to generate
+
+        Returns
+        -------
+        list of str or (str, str)
+            A list of str of string pairs (i5, i7) depending if the indexing
+            technology is dual barcoded or not
+
+        Notes
+        -----
+            This generates a list of i5 barcodes or the combination of
+            i5 and i7 barcodes. The combinations are more complex as we need
+            to make sure that we have actual rotation of those combinations.
+            Let's imagine we number each of the 384 unique i5 and i7 barcodes
+            1-384 (going across rows), in initial plate of 384 samples
+            `sample 1` will get i5_1 and i7_1, `sample 2` will get i5_2 and
+            i7_2, and so on. For the next plate of 384 samples, `sample 1`
+            will get i5_1 and i7_2, `sample 2` will get i5_2 and i7_3,
+            and so on. After 147,456 samples it will reset back to i5_1
+            and i7_1.
+        """
+        with TRN:
+            sql = """SELECT shotgun_index_tech_id, name, dual_index,
+                        last_index_idx
+                     FROM pm.shotgun_index_tech WHERE name = %s"""
+            TRN.add(sql, [idx_tech])
+            idx_tech_vals = TRN.execute_fetchindex()
+
+            if not idx_tech_vals:
+                raise ValueError('Not a valid shotgun index '
+                                 'technology: %s' % idx_tech)
+            else:
+                # just taking the first element and converting it to dict
+                idx_tech_vals = dict([v for v in idx_tech_vals][0])
+
+            # selecting single indexing
+            sql = """SELECT shotgun_index_id
+                     FROM pm.shotgun_index
+                     WHERE shotgun_index_tech_id = %s AND index_type = 'i5'
```

Comment:
Single index technologies are just i7, so this should be reversed -- only add i5 in dual index strategies.